builds/jenkins-build-pack-function/Jenkinsfile

```diff
@@ -37,376 +37,410 @@ import groovy.transform.Field
+                def role_name = "${configLoader.INSTANCE_PREFIX}-${config['domain']}-${config['service']}-${environment_logical_id}"
+                def role_arn = lambdaEvents.getRoleArn(role_name, credsId)
+                events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(config['provider'], role_arn, "iam_role", config), environment_logical_id);
+              }
 
-            if (domain != "jazz") {
-              createSubscriptionFilters(config, env_key, credsId, accountDetails);
-            }
+              if (domain != "jazz") {
+                createSubscriptionFilters(config, env_key, credsId, accountDetails);
+              }
 
-            if(!accountDetails.PRIMARY){
-              def primaryAccountData = utilModule.getAccountInfoPrimary();
-              def primaryAccountValue = primaryAccountData.ACCOUNTID;
-              def lambdaFnNameValue = "${configLoader.INSTANCE_PREFIX}-${config['domain']}-${config['service']}-${environment_logical_id}"
+              if(!accountDetails.PRIMARY){
+                def primaryAccountData = utilModule.getAccountInfoPrimary();
+                def primaryAccountValue = primaryAccountData.ACCOUNTID;
+                def lambdaFnNameValue = "${configLoader.INSTANCE_PREFIX}-${config['domain']}-${config['service']}-${environment_logical_id}"
 
-              addPermission(lambdaFnNameValue, primaryAccountValue, credsId, config);
-            }
+                addPermission(lambdaFnNameValue, primaryAccountValue, credsId, config);
+              }
+
+              if (domain == "jazz") {
+                serviceConfigdata.setLogStreamPermission(config)
+                serviceConfigdata.setKinesisStream(config)
+              }
 
-            if (domain == "jazz") {
-              serviceConfigdata.setLogStreamPermission(config)
-              serviceConfigdata.setKinesisStream(config)
+            } catch (ex) {
+              handleFail(ex, context_map, environment_logical_id, config, gitCommitHash)
+            } finally {
+              // reset Credentials
+              resetCredentials(credsId)
             }
+            environmentDeploymentMetadata.setEnvironmentEndpoint(lambdaARN)
+            def svc_response = echoServiceInfo(environment_logical_id)
+            sendFinalCompleteEvent(env_key, servicePlatform, environment_logical_id, config, gitCommitHash, gitCommitOwner, svc_response)
+          } //end of withCredentials
+        }
+      } //end of deployment to an environment
 
-          } catch (ex) {
-            handleFail(ex, context_map, environment_logical_id, config, gitCommitHash)
-          } finally {
-            // reset Credentials
-            resetCredentials(credsId)
+      //Udating safe details
+      if (config['domain'] != 'jazz' && configLoader.TVAULT && configLoader.TVAULT.IS_ENABLED) {
+        stage('Update Safe') {
```

Comment:
We will need to add roles to the safe before the deployment. Imagine RoleA is being used for a function. User supplied RoleB to be used for the function. If we deploy the service with RoleB and fail to add RoleB to the safe, function will now fail to access the secrets that it could access previously.

Reply:
Please see the explanation above